core/engine/src/builtins/intl/date_time_format/format.rs

```diff
@@ -0,0 +1,423 @@
+                era: None,
+                year: None,
+                month: None,
+                day: None,
+                day_period: None,
+                hour: None,
+                minute: None,
+                second: None,
+                fractional_second_digits: None,
+                time_zone_name: None,
+                ..
+            } => false,
+            // If any of the format fields is Some(_), return true
+            _ => true,
+        }
+    }
+
+    pub(super) fn check_dtf_type(&self, required: FormatType) -> bool {
+        // a. Let needDefaults be true.
+        // b. If required is date or any, then
+        // i. For each property name prop of « "weekday", "year", "month", "day" », do
+        // 1. Let value be formatOptions.[[<prop>]].
+        // 2. If value is not undefined, set needDefaults to false.
+        // c. If required is time or any, then
+        // i. For each property name prop of « "dayPeriod", "hour", "minute", "second", "fractionalSecondDigits" », do
+        // 1. Let value be formatOptions.[[<prop>]].
+        // 2. If value is not undefined, set needDefaults to false.
+        if required != FormatType::Time && self.has_date_defaults() {
+            return false;
+        }
+        if required != FormatType::Date && self.has_time_defaults() {
+            return false;
+        }
+        true
+    }
+
+    pub(super) fn has_date_defaults(&self) -> bool {
+        self.week_day.is_some() || self.year.is_some() || self.month.is_some() || self.day.is_some()
+    }
+
+    pub(super) fn has_time_defaults(&self) -> bool {
+        self.day_period.is_some()
+            || self.hour.is_some()
+            || self.minute.is_some()
+            || self.second.is_some()
+            || self.fractional_second_digits.is_some()
+    }
+
+    pub(super) fn to_length(&self) -> Option<Length> {
+        match (self.month, self.week_day) {
+            (Some(month), _) => Some(month.to_length()),
+            (None, Some(week_day)) => Some(week_day.to_length()),
+            _ => None,
+        }
+    }
+
+    /// Convert the current `FormatOptions` to a [`DateFields`].
+    pub(super) fn to_date_fields(&self) -> Option<DateFields> {
+        match (self.year, self.month, self.day, self.week_day) {
+            (Some(_y), _m, _d, Some(_e)) => Some(DateFields::YMDE),
+            (Some(_y), _m, Some(_d), None) => Some(DateFields::YMD),
+            (Some(_y), Some(_m), None, None) => Some(DateFields::YM),
+            (Some(_y), None, None, None) => Some(DateFields::Y),
+            (None, Some(_m), _d, Some(_e)) => Some(DateFields::MDE),
+            (None, Some(_m), Some(_d), None) => Some(DateFields::MD),
+            (None, Some(_m), None, None) => Some(DateFields::M),
+            (None, None, Some(_d), Some(_e)) => Some(DateFields::DE),
+            (None, None, Some(_d), None) => Some(DateFields::D),
+            (None, None, None, Some(_e)) => Some(DateFields::E),
+            _ => None,
```

Comment:
Suggestion: I think if you match on (None, None, None, None) then this match is exhaustive, which would be cool.